File: monitoring-system/server/app/services/monitoring_service.py
"""
Monitoring service — dashboard statistics and audit logs.
"""
from sqlalchemy import func
from ..extensions import db
from ..models.submission import Submission
from ..models.user import User
from ..models.assignment import AuditLog
from ..models.program import Program
from ..models.visitor_log import VisitorLog, VISITOR_STATUSES


def get_stats() -> tuple[dict, int]:
    """Return high-level dashboard statistics."""
    total         = Submission.query.count()
    pending       = Submission.query.filter_by(status="pending").count()
    under_review  = Submission.query.filter_by(status="under_review").count()
    assigned      = Submission.query.filter_by(status="assigned").count()
    resolved      = Submission.query.filter_by(status="resolved").count()
    closed        = Submission.query.filter_by(status="closed").count()
    total_users   = User.query.filter_by(role="user").count()
    secretaries   = User.query.filter_by(role="secretary", is_active=True).count()
    total_staff   = User.query.filter(User.role.in_(["admin", "secretary"])).count()

    # Submissions by category
    by_category = (
        db.session.query(Submission.category, func.count(Submission.id))
        .group_by(Submission.category)
        .all()
    )
    # Submissions by priority
    by_priority = (
        db.session.query(Submission.priority, func.count(Submission.id))
        .group_by(Submission.priority)
        .all()
    )
    # Last 7 days volume (daily counts)
    from datetime import datetime, timedelta, timezone
    today = datetime.now(timezone.utc).date()
    daily = []
    for i in range(6, -1, -1):
        day = today - timedelta(days=i)
        count = Submission.query.filter(
            func.date(Submission.created_at) == day
        ).count()
        daily.append({"date": day.isoformat(), "count": count})

    # Visitor log breakdown by status
    visitor_counts = {s: 0 for s in VISITOR_STATUSES}
    for status, cnt in (
        db.session.query(VisitorLog.status, func.count(VisitorLog.id))
        .group_by(VisitorLog.status)
        .all()
    ):
        visitor_counts[status] = cnt
    visitor_counts["total"] = sum(visitor_counts[s] for s in VISITOR_STATUSES)
    visitors_today = VisitorLog.query.filter(
        func.date(VisitorLog.created_at) == today
    ).count()
    visitor_counts["today"] = visitors_today

    return {
        "overview": {
            "total":        total,
            "pending":      pending,
            "under_review": under_review,
            "assigned":     assigned,
            "resolved":     resolved,
            "closed":       closed,
            "total_users":  total_users,
            "secretaries":  secretaries,
            "total_staff": total_staff,
        },
        "by_category": {cat: cnt for cat, cnt in by_category},
        "by_priority":  {pri: cnt for pri, cnt in by_priority},
        "daily_trend":  daily,
        "visitors": visitor_counts,
        "programs": {
            "active": Program.query.filter_by(is_active=True).count(),
            "total":  Program.query.count(),
        },
        "queue": {
            "oldest_pending": Submission.query.filter_by(status="pending")
                .order_by(Submission.created_at.asc(), Submission.id.asc()).first().to_dict()
                if Submission.query.filter_by(status="pending").first() else None,
        },
    }, 200


def get_audit_logs(page: int = 1, per_page: int = 50) -> tuple[dict, int]:
    paginated = (AuditLog.query
                 .order_by(AuditLog.timestamp.desc())
                 .paginate(page=page, per_page=per_page, error_out=False))
    return {
        "logs":     [log.to_dict() for log in paginated.items],
        "total":    paginated.total,
        "pages":    paginated.pages,
        "page":     page,
        "per_page": per_page,
    }, 200

File: monitoring-system/server/src/services/monitoringService.js
import { prisma } from '../db.js';
import { submissionDict, auditLogDict } from '../utils/serialize.js';

const VISITOR_STATUSES = ['pending', 'assigned', 'attended', 'closed'];

function dayRangeUTC(date) {
  const start = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()));
  const end = new Date(start.getTime() + 24 * 60 * 60 * 1000);
  return { start, end };
}

export async function getStats() {
  const [total, pending, under_review, assigned, resolved, closed] = await Promise.all([
    prisma.submissions.count(),
    prisma.submissions.count({ where: { status: 'pending' } }),
    prisma.submissions.count({ where: { status: 'under_review' } }),
    prisma.submissions.count({ where: { status: 'assigned' } }),
    prisma.submissions.count({ where: { status: 'resolved' } }),
    prisma.submissions.count({ where: { status: 'closed' } }),
  ]);

  const [total_users, secretaries, total_staff] = await Promise.all([
    prisma.users.count({ where: { role: 'user' } }),
    prisma.users.count({ where: { role: 'secretary', is_active: true } }),
    prisma.users.count({ where: { role: { in: ['admin', 'secretary'] } } }),
  ]);

  const catGroups = await prisma.submissions.groupBy({ by: ['category'], _count: { id: true } });
  const priGroups = await prisma.submissions.groupBy({ by: ['priority'], _count: { id: true } });
  const by_category = Object.fromEntries(catGroups.map((g) => [g.category, g._count.id]));
  const by_priority = Object.fromEntries(priGroups.map((g) => [g.priority, g._count.id]));

  const now = new Date();
  const daily = [];
  for (let i = 6; i >= 0; i -= 1) {
    const d = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate() - i));
    const { start, end } = dayRangeUTC(d);
    // eslint-disable-next-line no-await-in-loop
    const count = await prisma.submissions.count({ where: { created_at: { gte: start, lt: end } } });
    daily.push({ date: start.toISOString().slice(0, 10), count });
  }

  const vGroups = await prisma.visitor_logs.groupBy({ by: ['status'], _count: { id: true } });
  const visitors = Object.fromEntries(VISITOR_STATUSES.map((s) => [s, 0]));
  for (const g of vGroups) visitors[g.status] = g._count.id;
  visitors.total = VISITOR_STATUSES.reduce((sum, s) => sum + visitors[s], 0);
  const todayRange = dayRangeUTC(now);
  visitors.today = await prisma.visitor_logs.count({
    where: { created_at: { gte: todayRange.start, lt: todayRange.end } },
  });

  const [programsActive, programsTotal] = await Promise.all([
    prisma.programs.count({ where: { is_active: true } }),
    prisma.programs.count(),
  ]);

  const oldest = await prisma.submissions.findFirst({
    where: { status: 'pending' },
    orderBy: [{ created_at: 'asc' }, { id: 'asc' }],
    include: { submitter: true },
  });

  return {
    overview: {
      total, pending, under_review, assigned, resolved, closed,
      total_users, secretaries, total_staff,
    },
    by_category,
    by_priority,
    daily_trend: daily,
    visitors,
    programs: { active: programsActive, total: programsTotal },
    queue: { oldest_pending: oldest ? submissionDict(oldest) : null },
  };
}

export async function getAuditLogs(page = 1, perPage = 50) {
  const [total, items] = await prisma.$transaction([
    prisma.audit_logs.count(),
    prisma.audit_logs.findMany({
      include: { actor: true },
      orderBy: { timestamp: 'desc' },
      skip: (page - 1) * perPage,
      take: perPage,
    }),
  ]);
  return {
    logs: items.map(auditLogDict),
    total,
    pages: total === 0 ? 0 : Math.ceil(total / perPage),
    page,
    per_page: perPage,
  };
}
